aws-datastore/src/main/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapter.java

```diff
@@ -754,18 +756,30 @@ private synchronized void bindValueToStatement(
         }
     }
 
-    private boolean modelExists(Model model, QueryPredicate predicate) throws DataStoreException {
+    /**
+     * Helper method to synchronously query for a single model instance.  Used before saving and deleting to check if
+     * the item exists.
+     * @param model a Model that we want to query for the same type and id in SQLite.
+     * @param predicate a predicate to apply to the query.
+     * @return the Model instance from SQLite, if it exists, otherwise null.
+     */
+    private Model query(Model model, QueryPredicate predicate) {
         final String modelName = getModelName(model);
         final ModelSchema schema = modelSchemaRegistry.getModelSchemaForModelClass(modelName);
         final SQLiteTable table = SQLiteTable.fromSchema(schema);
-        final String tableName = table.getName();
         final String primaryKeyName = table.getPrimaryKeyColumnName();
 
         final QueryPredicate matchId = QueryField.field(primaryKeyName).eq(model.getId());
         final QueryPredicate condition = matchId.and(predicate);
-        try (Cursor cursor = getQueryAllCursor(tableName, Where.matches(condition))) {
-            return cursor.getCount() > 0;
-        }
+
+        Iterator<? extends Model> result = Single.<Iterator<? extends Model>>create(emitter -> {
+            if (model instanceof SerializedModel) {
+                query(modelName, Where.matches(condition), emitter::onSuccess, emitter::onError);
+            } else {
+                query(model.getClass(), Where.matches(condition), emitter::onSuccess, emitter::onError);
+            }
+        }).blockingGet();
+        return result.hasNext() ? result.next() : null;
     }
```

Comment:
the reason for implementing `modelExist()` method explicitly rather than using `query()` was to remove the workload required to serialize queried models when quickly checking if an item exists in the database. Not sure how much of a boost you get out of it though.

Reply:
That is a valid performance concern!  For the use case addressed by this PR, I need to serialize the queried models, in order to compute which fields have changed, for any save initiated by the client.  That said, this isn't necessary when saving models that come from the server, via the `SyncProcessor`, or `SubscriptionProcessor`.   We actually know this based on the `initiator` parameter of the `save` method.

Let me see what I can do to eliminate any unnecessary work based on that parameter.

Reply:
I just pushed a commit which reverts the changes I made to `modelExists()`, and adds a new `query` method, which is only called when we know this is an update mutation, _and_ it was initiated by `DATASTORE_API`.

This should ensure there won't be any negative performance impact on saves from the `SyncProcessor` or `SubscriptionProcessor`, or when creating an object via the DataStore API.